src/operation_cost.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashMap;
+
+use anyhow::{anyhow, Result};
+use apollo_compiler::{values::Selection, ApolloCompiler};
+
+use crate::compiler_ext::CompilerAdditions;
+
+struct Context<'a> {
+    compiler: &'a ApolloCompiler,
+    cost_map: &'a HashMap<String, i32>,
+}
+
+pub fn operation_cost(
+    sdl: &String,
+    operation: &String,
+    operation_name: Option<&String>,
+    cost_map: &HashMap<String, i32>,
+) -> Result<i32> {
+    let mut input = sdl.to_owned();
+    input.push_str(operation.as_str());
+
+    let compiler = ApolloCompiler::new(&input);
+
+    let context = Context {
+        compiler: &compiler,
+        cost_map,
+    };
+
+    if let Some(operation) = compiler.operation_by_name(operation_name) {
+        let parent = compiler
+            .operation_root_type(&operation)
+            .expect("root type must exist");
+
+        let total_cost = recurse_selections(
+            &context,
+            operation.selection_set().selection(),
+            &parent.name().to_string(),
+        );
+
+        return Ok(total_cost);
+    }
+
+    Err(anyhow!("missing operation"))
```

Comment:
Since you have to deal with both Some and None, I think it's neater to:
```suggestion
    match compiler.operation_by_name(operation_name) {
        Some(operation) => {
            let parent = compiler
                .operation_root_type(&operation)
                .expect("root type must exist");
            Ok(recurse_selections(
                &context,
                operation.selection_set().selection(),
                parent.name(),
            ))
        }
        None => Err(anyhow!("missing operation")),
    }
```